tinytuya/__init__.py

```diff
@@ -263,28 +263,55 @@ def __init__(self, dev_id, address, local_key="", dev_type="default", connection
         self.retry = True
         self.dev_type = dev_type
         self.port = 6668  # default - do not expect caller to pass in
+        self.socket=None
+
+    def __del__(self):
+        if self.socket!=None:
+            self.socket.close()
+            self.socket=None
 
     def __repr__(self):
         return '%r' % ((self.id, self.address),)  # FIXME can do better than this
 
+    def _get_socket(self,renew):
+        if(renew and self.socket!=None):
+            self.socket.close()
+            self.socket=None
+        if(self.socket==None):
+          self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+          #s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+          self.socket.settimeout(self.connection_timeout)
+          self.socket.connect((self.address, self.port))
+
     def _send_receive(self, payload):
         """
         Send single buffer `payload` and receive a single buffer.
         
         Args:
             payload(bytes): Data to send.
         """
-        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-        s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
-        s.settimeout(self.connection_timeout)
-        s.connect((self.address, self.port))
-        s.send(payload)
-        data = s.recv(1024)
-        # Some devices fail to send full payload in first response
-        if self.retry and len(data) < 40:  
+        #s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        #s.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+        #s.settimeout(self.connection_timeout)
+        #s.connect((self.address, self.port))
+        success=False
+        while not success:
+          # make sure I have a socket
+          self._get_socket(False)
+          try:
+            self.socket.send(payload)
+            data = self.socket.recv(1024)
+            # Some devices fail to send full payload in first response
+            if self.retry and len(data) < 40:  
+                time.sleep(0.1)
+                data = self.socket.recv(1024)  # try again
+            success=True
+            #s.close()
+          except:
+            print("exception with low level tinytua socket!!!.  will retry!!!")
             time.sleep(0.1)

```

Comment:
This should probably be a proper log message instead of print.  For debugging I like the print but let me know if you want me to make a branch with a log instead.